sky/data/skystorage.py

```diff
@@ -0,0 +1,227 @@
+    diff = interval - elapsed_time
+    if diff <= 0:
+        return 0
+    else:
+        return diff
+
+
+def is_locked(file_path: str):
+    """Checks if the given lock file is locked"""
+    locked = False
+    if os.path.exists(file_path):
+        lock = filelock.FileLock(file_path)
+        try:
+            lock.acquire(timeout=0)
+        except filelock.Timeout:
+            locked = True
+        finally:
+            if not locked:
+                lock.release()
+    return locked
+
+
+def set_s3_sync_cmd(src_path: str, bucketname: str, num_threads: int,
+                    delete: bool, no_follow_symlinks: bool):
+    """Builds sync command for aws s3"""
+    config_cmd = ('aws configure set default.s3.max_concurrent_requests '
+                  f'{num_threads}')
+    subprocess.check_output(config_cmd, shell=True)
+    sync_cmd = f'aws s3 sync {src_path} s3://{bucketname}'
+    if delete:
+        sync_cmd += ' --delete'
+    if no_follow_symlinks:
+        sync_cmd += ' --no-follow-symlinks'
+    return sync_cmd
+
+
+def set_gcs_sync_cmd(src_path: str, bucketname: str, num_threads: int,
+                     delete: bool, no_follow_symlinks: bool):
+    """Builds sync command for gcp gcs"""
+    sync_cmd = (f'gsutil -m -o \'GSUtil:parallel_thread_count={num_threads}\' '
+                'rsync -r')
+    if delete:
+        sync_cmd += ' -d'
+    if no_follow_symlinks:
+        sync_cmd += ' -e'
+    sync_cmd += f' {src_path} gs://{bucketname}'
+    return sync_cmd
+
+
+def run_sync(src: str,
+             storetype: str,
+             bucketname: str,
+             num_threads: int,
+             interval: int,
+             delete: bool,
+             no_follow_symlinks: bool,
+             max_retries: int = 10):
+    """
+    Runs the sync command to from SRC to STORETYPE bucket
+    """
+    #TODO: add enum type class to handle storetypes
+    storetype = storetype.lower()
+    if storetype == 's3':
+        sync_cmd = set_s3_sync_cmd(src, bucketname, num_threads, delete,
+                                   no_follow_symlinks)
+    elif storetype == 'gcs':
+        sync_cmd = set_gcs_sync_cmd(src, bucketname, num_threads, delete,
+                                    no_follow_symlinks)
+    else:
+        raise ValueError(f'Unknown store type: {storetype}')
+
+    log_file_name = f'csync_{storetype}_{bucketname}.log'
+    base_dir = os.path.expanduser(C_SYNC_FILE_PATH)
+    log_path = os.path.expanduser(os.path.join(base_dir, log_file_name))
+
+    with open(log_path, 'a') as fout:
+        try:
+            subprocess.run(sync_cmd,
+                           shell=True,
+                           check=True,
+                           stdout=fout,
+                           stderr=fout)
+        except subprocess.CalledProcessError:
+            if max_retries > 0:
+                #TODO: display the error with remaining # of retries
+                wait_time = interval / 2
```

Comment:
Why is the wait time set to `interval / 2`?

Reply:
Ideally, interval should be set to the amount of time that would take for the `sync` tool to upload the complete list of checkpoint, which allows to keep the storage and the node synced as much as possible without overlapped launch of the sync commands. For example, it took ~900ish seconds for `gsutil rsync` to sync 60GB of checkpoints from our original Vicuna example. So it would be nice to keep the interval as 900s in this case. 

As the error we are trying to handle here with the retry occurs when coincidentally training script is writing to the mounted directory and syncing from `gsutil rsync` occurs at the same time, it can take some time for this already running sync process to end. So I updated with a `Backoff` object from `common_utils` with `interval/2` of initial backoff.